### PR TITLE
Add support for LEFT/RIGHT/FULL/INNER lateral join

### PR DESCRIPTION
**Pull request by kasiafi**
relates to #9

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java [hunk 1]
```diff
@@ -227,12 +227,13 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
         // The subquery's EnforceSingleRowNode always produces a row, so the join is effectively INNER
-        return appendLateralJoin(subPlan, subqueryPlan, scalarSubquery.getQuery(), correlationAllowed, LateralJoinNode.Type.INNER);
+        return appendLateralJoin(subPlan, subqueryPlan, scalarSubquery.getQuery(), correlationAllowed, LateralJoinNode.Type.INNER, TRUE_LITERAL);
     }
 
-    public PlanBuilder appendLateralJoin(PlanBuilder subPlan, PlanBuilder subqueryPlan, Query query, boolean correlationAllowed, LateralJoinNode.Type type)
+    public PlanBuilder appendLateralJoin(PlanBuilder subPlan, PlanBuilder subqueryPlan, Query query, boolean correlationAllowed, LateralJoinNode.Type type, Expression filterCondition)
     {
         PlanNode subqueryNode = subqueryPlan.getRoot();
+
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: This formatting change is unrelated to the core of this commit, so please remove it or pull it out as a separate commit.

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java [hunk 2]
```diff
@@ -227,12 +227,13 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
         }
 
         // The subquery's EnforceSingleRowNode always produces a row, so the join is effectively INNER
-        return appendLateralJoin(subPlan, subqueryPlan, scalarSubquery.getQuery(), correlationAllowed, LateralJoinNode.Type.INNER);
+        return appendLateralJoin(subPlan, subqueryPlan, scalarSubquery.getQuery(), correlationAllowed, LateralJoinNode.Type.INNER, TRUE_LITERAL);
     }
 
-    public PlanBuilder appendLateralJoin(PlanBuilder subPlan, PlanBuilder subqueryPlan, Query query, boolean correlationAllowed, LateralJoinNode.Type type)
+    public PlanBuilder appendLateralJoin(PlanBuilder subPlan, PlanBuilder subqueryPlan, Query query, boolean correlationAllowed, LateralJoinNode.Type type, Expression filterCondition)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: Joins don't have "filter" conditions. Rather, it's a "join" condition or criteria. The distinction is subtle but important. A filter determines whether a row is preserved or removed, but the join condition determines whether the rows are considered as candidates to be joined. Depending on the type of join, rows that don't satisfy the join condition may be dropped or joined with a synthetic row containing nulls (e.g., for outer joins). So I'd rename this argument to just `criteria`.

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java [hunk 3]
```diff
@@ -499,7 +501,7 @@ private PlanBuilder createPlanBuilder(Node node)
     private Set<Expression> extractOuterColumnReferences(PlanNode planNode)
     {
         // at this point all the column references are already rewritten to SymbolReference
-        // when reference expression is not rewritten that means it cannot be satisfied within given PlaNode
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: Move this fix to a separate commit, since it's unrelated to this change.

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java [hunk 4]
```diff
@@ -63,6 +85,7 @@
      */
     private final List<Symbol> correlation;
     private final Type type;
+    private final Expression filter;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java]: This should be called `criteria` (see my other comment in this PR).

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -546,6 +556,52 @@ private RelationPlan planLateralJoin(Join join, RelationPlan leftPlan, Lateral l
...
     }
 
+    private TranslationMap filterExpressionTranslationMap(RelationPlan leftPlan, RelationPlan rightPlan, PlanBuilder leftPlanBuilder, PlanBuilder rightPlanBuilder, Join join)
+    {
+        List<Symbol> rewriterOutputSymbols = ImmutableList.<Symbol>builder()
+                .addAll(leftPlan.getFieldMappings())
+                .addAll(rightPlan.getFieldMappings())
+                .build();
+
+        PlanNode root = new JoinNode(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: This doesn't seem right. Planning a join + lateral should not involve a join node. After all, we're modeling it as a specialized "apply + join" operation. Can you explain why you think this is needed?

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 6]
```diff
@@ -546,6 +556,52 @@ private RelationPlan planLateralJoin(Join join, RelationPlan leftPlan, Lateral l
         return new RelationPlan(planBuilder.getRoot(), analysis.getScope(join), outputSymbols);
     }
 
+    private TranslationMap filterExpressionTranslationMap(RelationPlan leftPlan, RelationPlan rightPlan, PlanBuilder leftPlanBuilder, PlanBuilder rightPlanBuilder, Join join)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: I don't see much value in having this as a separate method. There's no reuse across various operations and it's tied to the specific implementation of planning join + lateral. I would inline it in the method that uses is and add some comments in the appropriate sections if there's any logical separation worth highlighting.

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 7]
```diff
@@ -546,6 +556,52 @@ private RelationPlan planLateralJoin(Join join, RelationPlan leftPlan, Lateral l
...
+
+        RelationPlan intermediateRootRelationPlan = new RelationPlan(root, analysis.getScope(join), rewriterOutputSymbols);
+        TranslationMap translationMap = new TranslationMap(intermediateRootRelationPlan, analysis, lambdaDeclarationToSymbolMap);
+        translationMap.setFieldMappings(rewriterOutputSymbols);
+        translationMap.putExpressionMappingsFrom(leftPlanBuilder.getTranslations());
+        translationMap.putExpressionMappingsFrom(rightPlanBuilder.getTranslations());
+        return translationMap;
+    }
+
+    private Expression extractFilterExpression(Join joinNode)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: Similarly, I'd inline this where it's being used. It's very specific to the planning of that type of join (i.e., join + lateral).

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java [hunk 8]
```diff
@@ -61,10 +64,23 @@ public Result apply(LateralJoinNode lateralJoinNode, Captures captures, Context
                         decorrelatedNode.getNode(),
                         ImmutableList.of(),
                         lateralJoinNode.getOutputSymbols(),
-                        decorrelatedNode.getCorrelatedPredicates(),
+                        conjunctFilters(decorrelatedNode.getCorrelatedPredicates(), lateralJoinNode.getFilter()),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java]: Use `ExpressionUtils.combineConjuncts()`:

```java
ExpressionUtils.combineConjuncts(
    decorrelatedNode.getCorrelatedPredicates().orElse(TRUE_LITERAL), 
    lateralJoinNode.getFilter());
```

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/JoinNode.java [hunk 9]
```diff
@@ -201,7 +201,6 @@ public String getJoinLabel()
 
         public static Type typeConvert(Join.Type joinType)
         {
-            // Omit SEMI join types because they must be inferred by the planner and not part of the SQL parse tree
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/JoinNode.java]: Unrelated change?

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java (hunk 4 above)
Comment (Praveen2112) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java]: Can we maintain it as `Optional<Expression>` like we maintain in `JoinNode` ?

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: It looks like a egg-chicken-egg problem :). We do the same for join. We create join node to rewrite expression to create join. As I understand it is required to create a `TranslationMap`, because it is needed by `RelationPlan`. However, `PlanNode` from `RelationPlan` is not used in `TranslationMap`. I think we should refactor that and remove a need of `PlanNode` for `TranslationMap`.

Also I think this code should belong `SubqueriesPlanner`. `SubqueriesPlanner` should do proper expression rewrites.

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java [hunk 10]
```diff
@@ -279,7 +281,7 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
 
         PlanNode subqueryPlanRoot = subqueryPlan.getRoot();
         if (isAggregationWithEmptyGroupBy(subqueryPlanRoot)) {
-            subPlan.getTranslations().put(existsPredicate, BooleanLiteral.TRUE_LITERAL);
+            subPlan.getTranslations().put(existsPredicate, TRUE_LITERAL);
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: separete commit

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java (hunk 8 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java]: Maybe `decorrelatedNode.getCorrelatedPredicates()` should return `TRUE_LITERAL` instead of `Optional.empty()`. I think having nested optional structure is not needed, because first we need to check if value is empty() and if it equal to `TRUE_LITERAL`.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 11]
```diff
@@ -812,11 +817,25 @@ public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Set<Symb
         @Override
         public PlanNode visitLateralJoin(LateralJoinNode node, RewriteContext<Set<Symbol>> context)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: nit: `context` here means dependency symbols. I would rename this in entire file as separate commit.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 12]
```diff
@@ -812,11 +817,25 @@ public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Set<Symb
...
-            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty() && isScalar(subquery)) {
-                return context.rewrite(node.getInput(), context.get());
+            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty()) {
+                // remove unused lateral subquery of inner join
+                if (node.getType() == INNER && isScalar(subquery) && node.getFilter().equals(TRUE_LITERAL)) {
+                    return context.rewrite(node.getInput(), context.get());
+                }
+                // remove unused lateral subquery of left join
+                if (node.getType() == LEFT && isAtMostScalar(subquery)) {
+                    return context.rewrite(node.getInput(), context.get());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: do you have tests for that? (you can run `TestSubqueries` with coverage and see if this line was executed, then I would like to see focused test for this.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 13]
```diff
@@ -812,11 +817,25 @@ public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Set<Symb
...
+                    .build();
+
+            PlanNode subquery = context.rewrite(node.getSubquery(), expectedFilterAndContextSymbols);
 
             // remove unused lateral nodes
-            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty() && isScalar(subquery)) {
-                return context.rewrite(node.getInput(), context.get());
+            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty()) {
+                // remove unused lateral subquery of inner join
+                if (node.getType() == INNER && isScalar(subquery) && node.getFilter().equals(TRUE_LITERAL)) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: we could also check if subquery cardinality is 0, then we could optimize it to empty `ValuesNode`.

Also I think we should extract this to optimizer Rule and remove this code from here.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 14]
```diff
@@ -825,18 +844,29 @@ public PlanNode visitLateralJoin(LateralJoinNode node, RewriteContext<Set<Symbol
...
                     .build();
             PlanNode input = context.rewrite(node.getInput(), inputContext);
 
-            // remove unused lateral nodes
-            if (intersection(ImmutableSet.copyOf(input.getOutputSymbols()), inputContext).isEmpty() && isScalar(input)) {
-                return subquery;
+            // remove unused input nodes
+            if (intersection(ImmutableSet.copyOf(input.getOutputSymbols()), expectedCorrelationAndContextSymbols).isEmpty()) {
+                // remove unused input of inner join
+                if (node.getType() == INNER && isScalar(input) && node.getFilter().equals(TRUE_LITERAL)) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: same comments here

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 15]
```diff
@@ -453,8 +453,9 @@ public PlanNode visitLateralJoin(LateralJoinNode node, RewriteContext<Void> cont
             PlanNode source = context.rewrite(node.getInput());
             PlanNode subquery = context.rewrite(node.getSubquery());
             List<Symbol> canonicalCorrelation = canonicalizeAndDistinct(node.getCorrelation());
+            Expression canonicalFilter = canonicalize(node.getFilter());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java]: inline

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/LateralJoinNode.java]: > Can we maintain it as Optional<Expression> like we maintain in JoinNode ?

I would prefer the opposite. To remove `Optional` from `JoinNode`. To me it is nested optional-like structure, two levels where you need to check for emptiness.

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/Patterns.java [hunk 16]
```diff
@@ -232,6 +232,11 @@ private Patterns() {}
         {
             return property("subquery", LateralJoinNode::getSubquery);
         }
+
+        public static Property<LateralJoinNode, Lookup, Expression> filter()
+        {
+            return property("filter", LateralJoinNode::getFilter);
+        }
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/Patterns.java]: you could also add pattern for empty critieria (equal to `TRUE_LITERAL`), as it looks to be used quite frequently.

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 5 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: @kokosing you're right. It's just a dummy used to create a `TranslationMap` and not used in the resulting plan. It was inspired by `RelationPlanner.visitJoin()`

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 6 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: I thought it would be better to graphically separate it, because it constructs the `JoinNode` which is *not* a part of the resulting plan (as mentioned above). Inlining the construction could be misleading at the first view.
Surely inlining + comment would be OK.

File: presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java (hunk 2 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java]: I'm aware of the naming bias.
However, `JoinNode` has got both members: `criteria` and `filter`.
I chose the name `filter` for `LateralJoinNode` member, because it is actually the counterpart of `JoinNode`'s `filter` and it becomes `JoinNode`'s `filter` later in the course of planning.

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 5 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: >  I think we should refactor that and remove a need of PlanNode for TranslationMap.

Let's not make structural changes quite yet. I have work in progress to rewrite RelationPlanner and friends so we can split the AST from the IR entirely, and it touches all that you're talking about. @kasiafi's addition is fine, though -- I can adapt it easily since it's an incremental change.

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java [hunk 17]
```diff
@@ -5041,16 +5041,21 @@ public void testLateralJoin()
...
+                "SELECT NULL, 'ALGERIA'");
+        assertQuery(
+                "SELECT * FROM (VALUES 1) a(x) FULL JOIN LATERAL(SELECT y FROM (VALUES 2) b(y) WHERE y > x) ON x=y",
+                "VALUES (1, NULL), (NULL, 2)");
+        assertQuery(
+                "SELECT * FROM (VALUES 1, 2, 3) a(x) FULL JOIN LATERAL(SELECT z FROM (VALUES 1, 2, 3, 5) b(z) WHERE z != x) ON x != 1 AND z != 5",
+                "VALUES (1, NULL), (2, 3), (2, 1), (3, 2), (3, 1), (NULL, 5)");
+        assertQuery(
+                "SELECT * FROM (VALUES 1, 2) a(x) JOIN LATERAL(SELECT y FROM (VALUES 2, 3) b(y) WHERE y > x) c(z) ON z > 2*x",
+                "VALUES (1, 3)");
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: Which tests exactly do you mean?

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java (hunk 12 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: There is already a test case which covers removing unused lateral subquery of inner join.
I'm adding another test case for left join.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java (hunk 14 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java]: I'm adding test cases for that.

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/Patterns.java (hunk 16 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/Patterns.java]: Could you show me some example I could follow?

Comment (Praveen2112) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/Patterns.java]: I guess we have to create something like this 
```
        public static Property<LateralJoinNode, Lookup, Boolean> isEmpty()
        {
            return property("isEmpty", lateralJoinNode -> lateralJoinNode.getFilter().equals(TRUE_LITERAL);
        }
```

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 5 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java]: Yes, apparently `ValuesNode` works fine and even with empty `rows`. That looks much better as a dummy, thanks.

**Comment by kasiafi** [comment]
@martint I applied your comments. Is there anything else to be done?

**Comment by martint** [comment]
Merged. Thanks for your contribution, @kasiafi!
